### PR TITLE
Transfer coremod to mixins

### PR DESCRIPTION
**Pull request by Alexdoru**
- Transfer (most of) the coremod to mixins, with the exception of one transformer
- Stop destructively overriding the player's InventoryPlayer field and instead inject new fields & methods inside the InventoryPlayer class. The old ASM was overwriting the inventory field in the EntityPlayer class and replacing it with a BattleGear inventory field.

- Stop destructively overriding the player's InventoryPlayer mainInventory array access and instead wrap it with an if statement.

The old ASM would replace the following vanilla code : 
```
if (itemstack.stackSize == 0) {
    this.thePlayer.inventory.mainInventory[this.thePlayer.inventory.currentItem] = null;
}
```
With :
```
if (itemstack.stackSize == 0) {
    BattlegearUtils.setPlayerCurrentItem(this.thePlayer, (ItemStack)null);
}
```
Now it changes it like this :
```
if (itemstack.stackSize == 0) {
    if (!InventoryAccessHook.setPlayerCurrentItem(this.thePlayer, (ItemStack)null)) {
        this.thePlayer.inventory.mainInventory[this.thePlayer.inventory.currentItem] = null;
    }
}
```

**Note: this commit contains breaking changes to the api, however it is only used in Tinkers constructs, see this other PR**

https://github.com/GTNewHorizons/TinkersConstruct/pull/99

### CODE REVIEW

**Comment by Dream-Master** [comment]
@Alexdoru can you setup this prs as draft ?

**Comment by Dream-Master** [comment]
@Alexdoru when the other changes are made in tico as a draft pr as well we can reopen it.

**Comment by Alexdoru** [comment]
> @Alexdoru when the other changes are made in tico as a draft pr as well we can reopen it.

but I need to use the new api in battlegear to make the changes in tico, how can I update the dependency in tico when this isn't merged ?

**Comment by Dream-Master** [comment]
i can make you a pre tag

File: gradle.properties [hunk 1]
```diff
@@ -15,46 +15,135 @@ autoUpdateBuildScript = false
...
+gradleTokenVersion = GRADLETOKEN_VERSION
+gradleTokenGroupName = GROUPNAME
+# [DEPRECATED]
+# Multiple source files can be defined here by providing a comma-seperated list: Class1.java,Class2.java,Class3.java
 # public static final String VERSION = "GRADLETOKEN_VERSION";
-# The string's content will be replaced with your mods version when compiled. You should use this to specify your mod's
+# The string's content will be replaced with your mod's version when compiled. You should use this to specify your mod's
 # version in @Mod([...], version = VERSION, [...])
 # Leave these properties empty to skip individual token replacements
 replaceGradleTokenInFile = Battlegear.java
```
Comment (eigenraven) [inline gradle.properties]:
```suggestion
replaceGradleTokenInFile =
```
Set this to an empty string, use either a generated Tags class or replacement, not both.

File: src/main/java/mods/battlegear2/asm/MethodMapping.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package mods.battlegear2.asm;
+
+import mods.battlegear2.asm.loader.BattlegearLoadingPlugin;
+
+public enum MethodMapping {
+
+    // spotless:off
+    ENTITYAICONTROLLEDBYPLAYER$UPDATETASK("e()V", "updateTask()V", 1),
```
Comment (eigenraven) [inline src/main/java/mods/battlegear2/asm/MethodMapping.java]: Can you add underscores here to separate the words? This is near-impossible to read

File: src/main/java/mods/battlegear2/asm/transformers/InventoryArrayAccessTransformer.java [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+                            }
+                            fourthNode = thirdNode.getNext().getNext();
+                        } else {
+                            fourthNode = thirdNode.getNext();
+                        }
+                        if (isLoadPlayerNode(fourthNode)) {
+                            final AbstractInsnNode fifthNode = fourthNode.getNext();
+                            if (isPlayerInventoryFieldNode(fifthNode)) {
+                                final AbstractInsnNode sixthNode = fifthNode.getNext();
+                                if (isCurrentItemFieldNode(sixthNode)) {
```
Comment (eigenraven) [inline src/main/java/mods/battlegear2/asm/transformers/InventoryArrayAccessTransformer.java:203]: This is very deeply nested, can you invert all the if conditions here and use `if(!isXYZ(node)) {continue;}`?

Comment (Alexdoru) [reply src/main/java/mods/battlegear2/asm/transformers/InventoryArrayAccessTransformer.java:203]: I know that I can do thatn and I tried it, but I personally prefer keeping the nested statements statements, I found that more readable than inverting everything and using continue

**Review by eigenraven — APPROVED** [review]

**Pull request merged**